File: compat/suites/go-sdk/internal/groups/stepfunctions.go
package groups

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sfn"
	sfntypes "github.com/aws/aws-sdk-go-v2/service/sfn/types"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/clients"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/harness"
)

func StepFunctions(c *clients.Clients) ServiceGroup {
	g := &sfnGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateStateMachine":               g.CreateStateMachine,
			"DescribeStateMachine":             g.DescribeStateMachine,
			"ListStateMachines":                g.ListStateMachines,
			"sfn-statemachines:StartExecution": g.StartExecution,
			"DeleteStateMachine":               g.DeleteStateMachine,

			// sfn-executions — group-qualified so StartExecution does not
			// collide with the sfn-statemachines test of the same name. This
			// group provisions its own state machine in setup.
			"sfn-executions:StartExecution":      g.ExecStartExecution,
			"sfn-executions:DescribeExecution":   g.DescribeExecution,
			"sfn-executions:GetExecutionHistory": g.GetExecutionHistory,
			"sfn-executions:ListExecutions":      g.ListExecutions,
			"sfn-executions:StopExecution":       g.StopExecution,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"sfn-statemachines": g.setupSFN,
			"sfn-executions":    g.setupExecutions,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"sfn-statemachines": g.teardownSFN,
			"sfn-executions":    g.teardownExecutions,
		},
	}
}

type sfnGroup struct{ c *clients.Clients }

func (g *sfnGroup) cl() *sfn.Client { return g.c.SFN() }

func (g *sfnGroup) setupSFN(_ context.Context, _ *harness.TestContext) error { return nil }

func (g *sfnGroup) teardownSFN(ctx context.Context, t *harness.TestContext) error {
	if arn := t.GetString("sfn_sm_arn"); arn != "" {
		g.cl().DeleteStateMachine(ctx, &sfn.DeleteStateMachineInput{StateMachineArn: aws.String(arn)}) //nolint:errcheck
	}
	return nil
}

func (g *sfnGroup) CreateStateMachine(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("compat-%s", t.RunID)
	def, _ := json.Marshal(map[string]interface{}{
		"Comment": "compat test",
		"StartAt": "Pass",
		"States": map[string]interface{}{
			"Pass": map[string]interface{}{"Type": "Pass", "End": true},
		},
	})
	roleArn := fmt.Sprintf("arn:aws:iam::000000000000:role/compat-sfn-%s", t.RunID)
	resp, err := g.cl().CreateStateMachine(ctx, &sfn.CreateStateMachineInput{
		Name:       aws.String(name),
		Type:       sfntypes.StateMachineTypeExpress,
		Definition: aws.String(string(def)),
		RoleArn:    aws.String(roleArn),
	})
	if err != nil {
		return err
	}
	if resp.StateMachineArn == nil {
		return fmt.Errorf("CreateStateMachine: missing stateMachineArn")
	}
	t.Set("sfn_sm_arn", *resp.StateMachineArn)
	return nil
}

func (g *sfnGroup) DescribeStateMachine(ctx context.Context, t *harness.TestContext) error {
	smArn := t.GetString("sfn_sm_arn")
	if smArn == "" {
		return fmt.Errorf("DescribeStateMachine: no state machine from CreateStateMachine")
	}
	resp, err := g.cl().DescribeStateMachine(ctx, &sfn.DescribeStateMachineInput{
		StateMachineArn: aws.String(smArn),
	})
	if err != nil {
		return err
	}
	if resp.StateMachineArn == nil {
		return fmt.Errorf("DescribeStateMachine: missing stateMachineArn")
	}
	return nil
}

func (g *sfnGroup) ListStateMachines(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().ListStateMachines(ctx, &sfn.ListStateMachinesInput{})
	return err
}

func (g *sfnGroup) StartExecution(ctx context.Context, t *harness.TestContext) error {
	smArn := t.GetString("sfn_sm_arn")
	if smArn == "" {
		return fmt.Errorf("StartExecution: no state machine from CreateStateMachine")
	}
	input, _ := json.Marshal(map[string]interface{}{"key": "value"})
	resp, err := g.cl().StartExecution(ctx, &sfn.StartExecutionInput{
		StateMachineArn: aws.String(smArn),
		Input:           aws.String(string(input)),
	})
	if err != nil {
		return err
	}
	if resp.ExecutionArn == nil {
		return fmt.Errorf("StartExecution: missing executionArn")
	}
	return nil
}

func (g *sfnGroup) DeleteStateMachine(ctx context.Context, t *harness.TestContext) error {
	smArn := t.GetString("sfn_sm_arn")
	if smArn == "" {
		return nil
	}
	_, err := g.cl().DeleteStateMachine(ctx, &sfn.DeleteStateMachineInput{
		StateMachineArn: aws.String(smArn),
	})
	return err
}

// ─── sfn-executions ───────────────────────────────────────────────────────────
//
// The execution engine really interprets the definition, so these tests assert
// on what the execution produced rather than just that a call succeeded.

const sfnExecDefinition = `{"Comment":"compat executions","StartAt":"Hello","States":{"Hello":{"Type":"Pass","Result":{"greeting":"hello"},"End":true}}}`

// setupExecutions provisions this group's own state machine — groups must not
// depend on resources another group created.
func (g *sfnGroup) setupExecutions(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("compat-exec-%s", t.RunID)
	resp, err := g.cl().CreateStateMachine(ctx, &sfn.CreateStateMachineInput{
		Name:       aws.String(name),
		Type:       sfntypes.StateMachineTypeExpress,
		Definition: aws.String(sfnExecDefinition),
		RoleArn:    aws.String(fmt.Sprintf("arn:aws:iam::000000000000:role/compat-sfn-exec-%s", t.RunID)),
	})
	if err != nil {
		return err
	}
	if resp.StateMachineArn == nil {
		return fmt.Errorf("setupExecutions: missing stateMachineArn")
	}
	t.Set("sfn_exec_sm_arn", *resp.StateMachineArn)
	return nil
}

func (g *sfnGroup) teardownExecutions(ctx context.Context, t *harness.TestContext) error {
	if arn := t.GetString("sfn_exec_sm_arn"); arn != "" {
		g.cl().DeleteStateMachine(ctx, &sfn.DeleteStateMachineInput{StateMachineArn: aws.String(arn)}) //nolint:errcheck
	}
	return nil
}

func (g *sfnGroup) ExecStartExecution(ctx context.Context, t *harness.TestContext) error {
	smArn := t.GetString("sfn_exec_sm_arn")
	if smArn == "" {
		return fmt.Errorf("StartExecution: no state machine from setup")
	}
	resp, err := g.cl().StartExecution(ctx, &sfn.StartExecutionInput{
		StateMachineArn: aws.String(smArn),
		Name:            aws.String(fmt.Sprintf("run-%s", t.RunID)),
		Input:           aws.String(`{"key":"value"}`),
	})
	if err != nil {
		return err
	}
	if resp.ExecutionArn == nil {
		return fmt.Errorf("StartExecution: missing executionArn")
	}
	t.Set("sfn_exec_arn", *resp.ExecutionArn)
	return nil
}

func (g *sfnGroup) DescribeExecution(ctx context.Context, t *harness.TestContext) error {
	execArn := t.GetString("sfn_exec_arn")
	if execArn == "" {
		return fmt.Errorf("DescribeExecution: no execution from StartExecution")
	}
	resp, err := g.cl().DescribeExecution(ctx, &sfn.DescribeExecutionInput{ExecutionArn: aws.String(execArn)})
	if err != nil {
		return err
	}
	if resp.Status == "" {
		return fmt.Errorf("DescribeExecution: missing status")
	}
	if resp.ExecutionArn == nil || *resp.ExecutionArn != execArn {
		return fmt.Errorf("DescribeExecution: executionArn did not round-trip")
	}
	return nil
}

func (g *sfnGroup) GetExecutionHistory(ctx context.Context, t *harness.TestContext) error {
	execArn := t.GetString("sfn_exec_arn")
	if execArn == "" {
		return fmt.Errorf("GetExecutionHistory: no execution from StartExecution")
	}
	resp, err := g.cl().GetExecutionHistory(ctx, &sfn.GetExecutionHistoryInput{ExecutionArn: aws.String(execArn)})
	if err != nil {
		return err
	}
	if len(resp.Events) == 0 {
		return fmt.Errorf("GetExecutionHistory: no events")
	}
	return nil
}

func (g *sfnGroup) ListExecutions(ctx context.Context, t *harness.TestContext) error {
	smArn := t.GetString("sfn_exec_sm_arn")
	if smArn == "" {
		return fmt.Errorf("ListExecutions: no state machine from setup")
	}
	resp, err := g.cl().ListExecutions(ctx, &sfn.ListExecutionsInput{StateMachineArn: aws.String(smArn)})
	if err != nil {
		return err
	}
	execArn := t.GetString("sfn_exec_arn")
	for _, item := range resp.Executions {
		if item.ExecutionArn != nil && *item.ExecutionArn == execArn {
			return nil
		}
	}
	return fmt.Errorf("ListExecutions: execution %s not listed", execArn)
}

func (g *sfnGroup) StopExecution(ctx context.Context, t *harness.TestContext) error {
	execArn := t.GetString("sfn_exec_arn")
	if execArn == "" {
		return fmt.Errorf("StopExecution: no execution from StartExecution")
	}
	_, err := g.cl().StopExecution(ctx, &sfn.StopExecutionInput{ExecutionArn: aws.String(execArn)})
	return err
}

File: compat/suites/python-sdk/groups/stepfunctions.py
"""
groups/stepfunctions.py — Step Functions compatibility test implementations for the Python suite.
"""

from __future__ import annotations
import json
from lib.harness import TestContext
from lib.clients import make_clients


def _sfn(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region)._get("stepfunctions")


# ── sfn-statemachines ─────────────────────────────────────────────────────────

def CreateStateMachine(ctx: TestContext) -> None:
    sfn = _sfn(ctx)
    name = f"compat-{ctx.run_id}"
    definition = json.dumps({
        "Comment": "compat test",
        "StartAt": "Pass",
        "States": {
            "Pass": {"Type": "Pass", "End": True},
        },
    })
    role_arn = f"arn:aws:iam::000000000000:role/compat-sfn-{ctx.run_id}"
    resp = sfn.create_state_machine(
        name=name,
        definition=definition,
        roleArn=role_arn,
        type="EXPRESS",
    )
    if not resp.get("stateMachineArn"):
        raise AssertionError("CreateStateMachine: missing stateMachineArn")
    ctx["sfn_sm_arn"] = resp["stateMachineArn"]


def DescribeStateMachine(ctx: TestContext) -> None:
    sfn = _sfn(ctx)
    sm_arn = ctx.get("sfn_sm_arn")
    if not sm_arn:
        raise AssertionError("DescribeStateMachine: no state machine from CreateStateMachine")
    resp = sfn.describe_state_machine(stateMachineArn=sm_arn)
    if not resp.get("stateMachineArn"):
        raise AssertionError("DescribeStateMachine: missing stateMachineArn")


def ListStateMachines(ctx: TestContext) -> None:
    _sfn(ctx).list_state_machines()


def StartExecution(ctx: TestContext) -> None:
    sfn = _sfn(ctx)
    sm_arn = ctx.get("sfn_sm_arn")
    if not sm_arn:
        raise AssertionError("StartExecution: no state machine from CreateStateMachine")
    resp = sfn.start_execution(
        stateMachineArn=sm_arn,
        input=json.dumps({"key": "value"}),
    )
    if not resp.get("executionArn"):
        raise AssertionError("StartExecution: missing executionArn")


def DeleteStateMachine(ctx: TestContext) -> None:
    sfn = _sfn(ctx)
    sm_arn = ctx.get("sfn_sm_arn")
    if not sm_arn:
        return
    sfn.delete_state_machine(stateMachineArn=sm_arn)


# ── sfn-executions ────────────────────────────────────────────────────────────
#
# The execution engine really interprets the definition, so these tests assert
# on what the execution produced rather than just that a call succeeded.

_EXEC_DEFINITION = json.dumps({
    "Comment": "compat executions",
    "StartAt": "Hello",
    "States": {"Hello": {"Type": "Pass", "Result": {"greeting": "hello"}, "End": True}},
})


def _setup_executions(ctx: TestContext) -> None:
    """Provision this group's own state machine — groups never share resources."""
    resp = _sfn(ctx).create_state_machine(
        name=f"compat-exec-{ctx.run_id}",
        definition=_EXEC_DEFINITION,
        roleArn=f"arn:aws:iam::000000000000:role/compat-sfn-exec-{ctx.run_id}",
        type="EXPRESS",
    )
    if not resp.get("stateMachineArn"):
        raise AssertionError("setup sfn-executions: missing stateMachineArn")
    ctx["sfn_exec_sm_arn"] = resp["stateMachineArn"]


def _teardown_executions(ctx: TestContext) -> None:
    sm_arn = ctx.get("sfn_exec_sm_arn")
    if sm_arn:
        try:
            _sfn(ctx).delete_state_machine(stateMachineArn=sm_arn)
        except Exception:
            pass


def ExecStartExecution(ctx: TestContext) -> None:
    sm_arn = ctx.get("sfn_exec_sm_arn")
    if not sm_arn:
        raise AssertionError("StartExecution: no state machine from setup")
    resp = _sfn(ctx).start_execution(
        stateMachineArn=sm_arn,
        name=f"run-{ctx.run_id}",
        input=json.dumps({"key": "value"}),
    )
    if not resp.get("executionArn"):
        raise AssertionError("StartExecution: missing executionArn")
    ctx["sfn_exec_arn"] = resp["executionArn"]


def DescribeExecution(ctx: TestContext) -> None:
    exec_arn = ctx.get("sfn_exec_arn")
    if not exec_arn:
        raise AssertionError("DescribeExecution: no execution from StartExecution")
    resp = _sfn(ctx).describe_execution(executionArn=exec_arn)
    if not resp.get("status"):
        raise AssertionError("DescribeExecution: missing status")
    if resp.get("executionArn") != exec_arn:
        raise AssertionError("DescribeExecution: executionArn did not round-trip")


def GetExecutionHistory(ctx: TestContext) -> None:
    exec_arn = ctx.get("sfn_exec_arn")
    if not exec_arn:
        raise AssertionError("GetExecutionHistory: no execution from StartExecution")
    resp = _sfn(ctx).get_execution_history(executionArn=exec_arn)
    if not resp.get("events"):
        raise AssertionError("GetExecutionHistory: no events")


def ListExecutions(ctx: TestContext) -> None:
    sm_arn = ctx.get("sfn_exec_sm_arn")
    if not sm_arn:
        raise AssertionError("ListExecutions: no state machine from setup")
    resp = _sfn(ctx).list_executions(stateMachineArn=sm_arn)
    exec_arn = ctx.get("sfn_exec_arn")
    arns = [e.get("executionArn") for e in resp.get("executions", [])]
    if exec_arn not in arns:
        raise AssertionError(f"ListExecutions: execution {exec_arn} not listed")


def StopExecution(ctx: TestContext) -> None:
    exec_arn = ctx.get("sfn_exec_arn")
    if not exec_arn:
        raise AssertionError("StopExecution: no execution from StartExecution")
    _sfn(ctx).stop_execution(executionArn=exec_arn)


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "sfn-statemachines:CreateStateMachine": CreateStateMachine,
    "sfn-statemachines:DescribeStateMachine": DescribeStateMachine,
    "sfn-statemachines:ListStateMachines": ListStateMachines,
    "sfn-statemachines:StartExecution": StartExecution,
    "sfn-statemachines:DeleteStateMachine": DeleteStateMachine,
    # sfn-executions — group-qualified so StartExecution does not collide with
    # the sfn-statemachines test of the same name.
    "sfn-executions:StartExecution": ExecStartExecution,
    "sfn-executions:DescribeExecution": DescribeExecution,
    "sfn-executions:GetExecutionHistory": GetExecutionHistory,
    "sfn-executions:ListExecutions": ListExecutions,
    "sfn-executions:StopExecution": StopExecution,
}

SETUP = {
    "sfn-executions": _setup_executions,
}
TEARDOWN = {
    "sfn-statemachines": lambda ctx: _teardown_state_machine(ctx),
    "sfn-executions": _teardown_executions,
}


def _teardown_state_machine(ctx: TestContext) -> None:
    sm_arn = ctx.get("sfn_sm_arn")
    if sm_arn:
        try:
            _sfn(ctx).delete_state_machine(stateMachineArn=sm_arn)
        except Exception:
            pass
